crates/sui-graphql-rpc/schema.graphql

```diff
@@ -0,0 +1,840 @@
+  signatures: [TransactionSignature]
+  effects: TransactionBlockEffects
+
+  expiration: Epoch
+
+  bcs: Bytes
+}
+
+enum TransactionBlockKindInput {
+  CONSENSUS_COMMIT_PROLOGUE
+  GENESIS
+  CHANGE_EPOCH
+  PROGRAMMABLE
+}
+
+union TransactionBlockKind =
+    ConsensusCommitPrologueTransaction
+  | GenesisTransaction
+  | ChangeEpochTransaction
+  | ProgrammableTransactionBlock
+
+type ConsensusCommitPrologueTransaction {
+  epoch: Epoch
+  # TODO: This is the "leader round" -- does this line up with
+  # checkpoints? In which case, it may suffice to have a `Checkpoint`
+  # here.
+  round: U64
+  timestampMs: U64
+}
+
+type GenesisTransaction {
+  objects: [Object]
+}
+
+type ChangeEpochTransaction {
+  epoch: Epoch
+  timestampMs: U64
+
+  storageCharge: U64
+  computationCharge: U64
+  storageRebate: U64
+}
+
+type ProgrammableTransactionBlock {
+  inputs: [TransactionInput]
+  transactions: [ProgrammableTransaction]
+}
+
+union TransactionInput = MovePackage | MoveObject | ShareInput
+
+type SharedInput {
+  id: SuiAddress
+  initialSharedVersion: U64
+  mutable: Boolean
+}
+
+union TransactionArgument = GasCoin | Input | Result
+
+type GasCoin {}
+type Input   { ix: Int }
+type Result  { cmd: Int, ix: Int }
+
+union ProgrammableTransaction =
+    MoveCallTransaction
+  | TransferObjectsTransaction
+  | SplitCoinTransaction
+  | MergeCoinsTransaction
+  | PublishTransaction
+  | UpgradeTransaction
+  | MakeMoveVecTransaction
+
+type MoveCallTransaction {
+  moduleId: MoveModuleId
+  name: String
+  function: NormalizedFunction
+
+  typeArguments: [MoveType]
+  arguments: [TransactionArgument]
+}
+
+type TransferObjectsTransaction {
+  objects: [TransactionArgument]
+  address: TransactionArgument
+}
+
+type SplitCoinTransaction {
+  coin: TransactionArgument
+  amounts: [TransactionArgument]
+}
+
+type MergeCoinsTransaction {
+  coin: TransactionArgument
+  coins: [TransactionArgument]
+}
+
+type PublishTransaction {
+  dependencies: [MovePackage]
+}
+
+type UpgradeTransaction {
+  dependencies: [MovePackage]
+  currentPackage: MovePackage
+  upgradeTicket: TransactionArgument
+}
+
+type MakeMoveVecTransaction {
+  type: MoveType
+  elements: [TransactionArgument]
+}
+
+# Omitting for brevity -- there's a lot of structure here, but it's
+# not very "connected" (i.e. no edges to other types).
+type TransactionSignature {}
+
+type TransactionBlockEffects {
+  digest: Digest!
+  status: ExecutionStatus!
+  errors: String
+  transactionBlock: TransactionBlock
+  dependencies: [TransactionBlock]
+
+  lamportVersion: U64
+  gasEffects: GasEffects
+  objectReads: [Object]
+  objectChanges: [ObjectChange]
+  balanceChanges: [BalanceChange]
+
+  epoch: Epoch
+  checkpoint: Checkpoint
+
+  events: TransactionEventConnection
+
+  bcs: Bytes
+}
+
+enum ExecutionStatus {
+  SUCCESS
+  FAILURE
+}
+
+type GasInput {
+  gasSponsor: Address
+  gasPayment: [Object]
+
+  gasPrice: U64
+  gasBudget: U64
+}
+
+type GasEffects {
+  gasObject: Coin
```

Comment:
Specialising the `gasObject` to `Coin`, because we know that's what it should be (we couldn't do the same for `gasPayment` because a failing transaction may have been passed bad gas objects).